code/crates/test/app/src/state.rs

```diff
@@ -520,33 +528,64 @@ impl State {
             .filter_map(|part| part.as_data())
             .fold(1, |acc, data| acc * data.factor);
 
+        // merge all the parts' extensions in order
+        let extensions: Bytes = parts
+            .parts
+            .iter()
+            .filter_map(|part| part.as_data())
+            .fold(BytesMut::new(), |mut acc, data| {
+                acc.extend_from_slice(&data.extension);
+                acc
+            })
+            .freeze();
+
         Ok(ProposedValue {
             height: parts.height,
             round: parts.round,
             valid_round: init.pol_round,
             proposer: parts.proposer,
-            value: Value::new(value),
+            value: Value::with_extensions(value, extensions),
             validity: Validity::Valid,
         })
     }
 }
 
 /// Encode a value to its byte representation
 pub fn encode_value(value: &Value) -> Bytes {
-    JsonCodec.encode(value).unwrap()
+    ProtobufCodec.encode(value).unwrap() // FIXME: unwrap
 }
 
 /// Decodes a Value from its byte representation
 pub fn decode_value(bytes: Bytes) -> Option<Value> {
-    JsonCodec.decode(bytes).ok()
+    ProtobufCodec.decode(bytes).ok()
+}
+
+/// Split `bytes` into `n` roughly-equal chunks
+fn split_bytes(mut bytes: Bytes, chunks: usize) -> Vec<Bytes> {
+    let total = bytes.len();
+    let chunk_len = total.div_ceil(chunks); // number of bytes a chunk can have
+
+    let mut out = Vec::with_capacity(chunks);
+    while out.len() < chunks {
+        if bytes.len() >= chunk_len {
+            out.push(bytes.split_to(chunk_len));
+        } else if !bytes.is_empty() {
+            out.push(bytes.split_to(bytes.len()));
+        } else {
+            out.push(Bytes::new());
+        }
+    }
+    out
 }
```

Comment:
I think this can be written better

```rs
fn split_bytes(mut bytes: Bytes, chunks: usize) -> Vec<Bytes> {
    if chunks == 0 {
        return vec![];
    }

    let total = bytes.len();
    let chunk_len = total.div_ceil(chunks);

    (0..chunks)
        .map(|_| {
            let split_size = cmp::min(bytes.len(), chunk_len);
            bytes.split_to(split_size)
        })
        .collect()
}
```